### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by chrisforrette**
#### What's this PR do?

Creates the first version of a Joi image extension, which validates image dimensions and types, and includes testing, coverage, linting, and publishing configuration.

@randallagordon One particular thing I'd like your input on is license—since this is our first open source thing and I know nothing about licenses and you do, let me know what would be best to fill in for the license in `package.json` and whether we might want to add any other indicators around that to the README or the repo.

#### Where should the reviewer start?

`src/index.js`

#### How should this be manually tested?

You can recreate the example from the README or in the unit tests if you want to try it out on your own image(s).

#### Any background context you want to provide?

For avatar upload we need to validate a minimum size so we're not stretching out avatar images, so I whipped this here Joi extension up.

#### What are the relevant tickets?

DOI-361

#### What gif best describes this PR or how it makes you feel?

https://giphy.com/gifs/mrw-dont-length-L1ZqGzJDFYC7C

#### Definition of Done:

- [x] Is there appropriate test coverage? (~80% coverage on all new code)
- [x] Is the README appropriately updated?
- [x] Are there breaking changes? Has a representative from each affected team signed-off?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,99 @@
+{
+  "name": "joi-image-extensions",
+  "version": "0.1.0",
```
Comment (randallagordon) [inline package.json]: Looks solid enough to jump straight to v1.0.0 to me!

File: package.json [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    "watch:mocha": "nodemon --exec 'npm run -s mocha' -w src -w test",
+    "watch:test": "nodemon --exec 'npm run -s test' -w src -w test"
+  },
+  "author": "chris@doing.io",
+  "license": "ISC",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/doinginc/joi-image-extension.git"
+  },
+  "license": "UNLICENSED",
```
Comment (randallagordon) [inline package.json]: I'd personally go with an MIT license, though ultimately we should get @technojosh's input since its a legal question. Though it likely doesn't make sense for this package since it's associated with Joi, any future work that may produce trademarks worth protecting warrants considering a BSD 3/4-clause.

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      maxWidth: 'must be no more than {{maxWidth}} pixels in width',
+      minHeight: 'must be at least {{minHeight}} pixels in height',
+      maxHeight: 'must be no more than {{maxHeight}} pixels in height',
+      minDimensions: 'must be at least {{minWidth}} pixels in width and {{minHeight}} pixels in height',
+      maxDimensions: 'must be no more than {{minWidth}} pixels in width and {{minHeight}} pixels in height',
+      allowTypes: 'must be one of the following types: {{allowedTypes}}',
+      disallowTypes: 'must be one of the following types: {{allowedTypes}}'
+    },
+
+    coerce (value, state, options) {
```
Comment (randallagordon) [inline src/index.js]: Would it be more appropriate to use `pre` here instead of `coerce`? I seem to recall there being a way/various reasons to turn off coercion and I think `pre` would still run regardless. Mainly thinking for chaining purposes where something else may need to turn off coercion for some reason, say an input form or CLI that specifies dimensions.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+#### What gif best describes this PR or how it makes you feel?
+
+
+
+#### Definition of Done:
+
+- [ ] Is there appropriate test coverage? (~80% coverage)
+- [ ] Is the README appropriately updated?
+- [ ] Are there breaking changes? Has a representative from each affected team signed-off?
```
Comment (randallagordon) [inline .github/PULL_REQUEST_TEMPLATE.md]: The wording here is a bit odd for an open source project, maybe drop the second question in favor of a statement such as "Please add notes about what they are and why the change is necessary."?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,89 @@
...
+    Joi.validate(image, schema, (err, value) => {
+      // Handle validation result...
+    })
+  }))
+
+```
+
+## API
+
+### `image.minWidth(minimum pixel width)`
```
Comment (randallagordon) [inline README.md:42]: When looking at the README as rendered by GitHub rather than as Markdown, this formatting looks a bit odd as the backtick usage implies a function signature. Maybe camelcase or just shorten them since the description immediately after explains it, `image.minWidth(minimumPixelWidth)`/`image.minWidth(minimum)`?

**Review by randallagordon — CHANGES_REQUESTED** [review]
This is lookin' rad! 💥  Just a handful of house-clean-y kinds of things.

File: README.md (hunk 5 above)
Comment (chrisforrette) [reply README.md:42]: @randallagordon I agree, it's kinda weird, but I'm just following suit with how Joi documents its stuff, which seems to be `(namespace).(function)`, e.g.: https://github.com/hapijs/joi/blob/v13.1.0/API.md#anyallowvalue

File: package.json (hunk 1 above)
Comment (chrisforrette) [reply package.json]: After this gets merged I was going to just run an `npm version` to do just that.

File: src/index.js (hunk 3 above)
Comment (chrisforrette) [reply src/index.js]: @randallagordon Indeed, it seems that way—good call.

**Comment by chrisforrette** [comment]
@randallagordon OK this should be all set—I've addressed all of your comments and I went ahead and just added an MIT license. The CircleCI stuff might need to be jiggered with once I get a tag flowing into it, and I think this PR is not running on it because I didn't enable it until after opening this.

File: README.md (hunk 5 above)
Comment (randallagordon) [reply README.md:42]: They don't break them up with spaces such that it is displaying invalid code, however:

image

vs:

image
